### DISCUSSION TITLE
how to build a dependency package for Skip?

### DISCUSSION THREAD

**Discussion by DevboiDesigns** [skiptools/skip#292] [Q&A] [answered]
Is there documentation any where on how to build a skip dependency? I have been looking but cannot find.

**Comment by marcprux** [comment, marked as answer]
What do you mean by dependency? Like a Skip package that depends on another Skip package, or one that depends on a pure Swift package on the iOS side and a Kotlin dependency on the Android side? That would be covered at https://skip.tools/docs/dependencies/, but let us know if you'd like clarification…

**Reply by DevboiDesigns** [reply]
My apologies, I should have clarified that better. This is what I was looking for.

_that depends on a pure Swift package on the iOS side and a Kotlin dependency on the Android side_

I want to work on creating a wrapper for the OneSignal API. 

I have been looking at the Firebase wrapper you have for ideas. I am fairly new at all this but can really see how amazing Skip is and am about to start a personal project with it. 

Thank you for your time 🙏

**Reply by marcprux** [reply]
I'm not familiar with the OneSignal API, but the general strategy we've found that works well with projects like [skip-firebase](https://github.com/skiptools/skip-firebase) and [skip-supabase](https://github.com/skiptools/skip-supabase) is to create an "adaptor" in Skip that mimics the Swift API of the framework, but is backed by the Kotlin implementation. That reduces the amount of work needed, since the well-tested Swift API for the functionality can be used directly. The only challenge is to coerce the Swift API into the Kotlin API.

In some cases like Firebase, whose [iOS](https://github.com/firebase/firebase-ios-sdk.git) and [Android](https://github.com/firebase/firebase-android-sdk.git) APIs are basically identical, this can be a fairly rote exercise. Other cases, like [skip-supabase](https://github.com/skiptools/skip-supabase), are more challenging because their [Swift](http://github.com/supabase/supabase-swift) and [Kotlin](https://github.com/supabase-community/supabase-kt) APIs are quite dissimilar.

Feel free to seek guidance, either here or on the [Slack Channel](https://skip.tools/slack), to help you implement this. We'd definitely be interested in seeing it added as part of the Skip ecosystem!

**Reply by DevboiDesigns** [reply]
Great! Thank you for all this detailed information. I joined the Slack for now, will reach out if I need to. Have a great day.
